sys/linux/vusb.txt

```diff
@@ -0,0 +1,749 @@
+# Copyright 2019 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+include <linux/types.h>
+include <linux/byteorder/generic.h>
+
+include <uapi/linux/usb/ch9.h>
+include <uapi/linux/usb/ch11.h>
+include <uapi/linux/hid.h>
+include <linux/hid.h>
+include <uapi/linux/usb/cdc.h>
+
+resource fd_usb[fd]
+
+syz_usb_connect(speed flags[usb_device_speed], dev_len len[dev], dev ptr[in, usb_device_descriptor], dev_id ptr[in, usb_device_id], conn_descs ptr[in, vusb_connect_descriptors]) fd_usb
+syz_usb_control_io(fd fd_usb, descs ptr[in, vusb_descriptors], resps ptr[in, vusb_responses])
+syz_usb_ep_write(fd fd_usb, ep int16[0:31], len len[data], data buffer[in])
+syz_usb_disconnect(fd fd_usb)
+
+usb_device_speed = USB_SPEED_UNKNOWN, USB_SPEED_LOW, USB_SPEED_FULL, USB_SPEED_HIGH, USB_SPEED_WIRELESS, USB_SPEED_SUPER, USB_SPEED_SUPER_PLUS
+
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+
+# USB device, configuration, interface and endpoint descriptors.
+
+usb_device_descriptor {
+	bLength			const[USB_DT_DEVICE_SIZE, int8]
+	bDescriptorType		const[USB_DT_DEVICE, int8]
+
+	bcdUSB			int16
+	bDeviceClass_p		flags[usb_classes, int8]
+	bDeviceSubClass_p	int8
+	bDeviceProtocol_p	int8
+	bMaxPacketSize0		int8
+	idVendor_p		int16
+	idProduct_p		int16
+	bcdDevice_p		int16
+	iManufacturer		int8
+	iProduct		int8
+	iSerialNumber		int8
+	bNumConfigurations	int8
+
+	configs			array[usb_config_descriptor, 1]
```

Comment:
Please add a comment as to why we use array of size 1. Do we need it?

Reply:
Added a comment above the struct.